flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/SinkCodeGenerator.scala

```diff
@@ -108,8 +108,7 @@ object SinkCodeGenerator {
     val inputTerm = CodeGenUtils.DEFAULT_INPUT1_TERM
     var afterIndexModify = inputTerm
     val fieldIndexProcessCode =
-      if (getCompositeTypes(convertOutputType).map(fromTypeInfoToLogicalType) sameElements
-          inputTypeInfo.getFieldTypes.map(fromTypeInfoToLogicalType)) {
+      if (getCompositeTypes(convertOutputType) sameElements inputTypeInfo.getFieldTypes) {
```

Comment:
it not work now...
`inputTypeInfo` is a BaseRowTypeInfo with internal type informations.
And `convertOutputType` is the type information from sink and users.
Consider Types.STRING and BinaryStringTypeInfo, Types.DECIMAL and DecimalTypeInfo.

Reply:
And I think maybe you can try `LogicalType.asSerializableString`